pkg/agent/openflow/client.go

```diff
@@ -101,26 +101,16 @@ func (c *client) IsConnected() bool {
 	return c.bridge.IsConnected()
 }
 
-// addMissingFlows adds any flow from flows which is not currently in the flow cache. The function
-// returns immediately in case of error when adding a flow. If a flow is added successfully, it is
-// added to the flow cache. If the flow cache has not been initialized yet (i.e. there is no
-// flowCacheKey key in the cache map), we create it first.
-func (c *client) addMissingFlows(cache *flowCategoryCache, flowCacheKey string, flows []binding.Flow) error {
-	// initialize flow cache if needed
+// addFlowsToCache adds the successfully installed flows into the flow cache. If the flow cache has not been initialized
+// yet (i.e. there is no flowCacheKey key in the cache map), we create it first.
+func (c *client) addFlowsToCache(cache *flowCategoryCache, flowCacheKey string, flows []binding.Flow) {
```

Comment:
@wenyingd are the comments for the `Client` interface methods still valid?
In particular at the moment the comments say that calls to `InstallNodeFlows` and `Install PodFlows` are idempotent.
I guess my question is really about the behavior of `AddFlowsInBundle` if some flows already exist. If some flows already exist, do the call to `AddFlowsInBundle` succeeds and is the following code executed unconditionally:
```
for _, flow := range addflows {
    ofFlow := flow.(*ofFlow)
    ofFlow.table.UpdateStatus(1)
    ofFlow.UpdateInstallStatus(true)
}
```
If yes, then the comment about the Install methods being idempotent is not valid.

Reply:
I don't think the assumption exists, for bundle could ensure flows in the same bundle are committed or discarded together, and no partially successful cases. Table status is changed only if all the flows are committed successfully, and the flows are added into the cache only after the `AddFlowsInBundle` returns nil. So the `InstallNodeFlows` and `InstallPodFlows` are still idempotent.

Reply:
Thanks @wenyingd. That answers my question. I would only suggest updating the comments for `InstallNodeFlows` and `InstallPodFlows`. For example, instead of just:
> Calls to InstallNodeFlows are idempotent.

We could have:
> InstallNodeFlows has all-or-nothing semantics (call succeeds iff all the flows are installed successfully, otherwise no flows will be installed). Calls to InstallNodeFlows are idempotent, but calling InstallNodeFlows again after a successful call will return an error, as all the flows already exist.

Same for `InstallPodFlows`. What do you think?

Reply:
Actually I think I may have misunderstood.
Let's say that for any reason, the client calls `InstallNodeFlows` twice with the exact set of parameters. If both calls succeed and `UpdateStatus` is called each time for each flow, then we end up with an incorrect flow count, and the call is not idempotent since the status is updated twice instead of once.
Your point is that the caller should *not* call `InstallNodeFlows` twice like this. Personally I'd rather have the Client API be more robust by: 1) keeping it idempotent and 2) returning an error if the flows have already been installed successfully. It seems to me that this can easily be achieved thanks to the flow cache. For example, in pseudo-code:
```
InstallNodeFlows {
  * check if flows for the Node are in cache
      * if yes, return an error
      * if no, add all the flows
          * if transaction fails, return an error
          * if transaction succeeds, update the flow cache for the Node and return nil
}
```

Reply:
Hi @wenyingd, I was thinking more along the lines of something like this:
```
func (c *client) addFlows(cache *flowCategoryCache, flowCacheKey string, flows []binding.Flow) error {
    if _, ok := cache.Load(flowCacheKey); ok {
        return fmt.Errorf("flows already installed")
    }
    if err := c.flowOperations.AddAll(flows); err != nil {
        return err
    }
    cache.Store(flowCacheKey, flows)
    return nil
}
func (c *client) InstallNodeFlows(hostname string, ...) error {
    // ...
    return addFlows(...)
}
```
I don't think there is a need to look at "missing flows" since partial success is not possible.

Reply:
Updated.